R/utils.R

```diff
@@ -367,3 +367,40 @@ set_num_threads <- function(num_threads) {
          call. = FALSE)
   }
 }
+
+check_divergences <- function(data_csv) {
+  divergences <- posterior::extract_one_variable_matrix(data_csv$sampler, "divergent__")
+  num_of_divergences <- sum(divergences)
+  if (num_of_divergences > 0) {
+    if(data_csv$sampling_info$sample_save_warmup == 1) {
+      num_iter <- data_csv$sampling_info$sample_num_samples + data_csv$sampling_info$sample_num_warmup
+    } else {
+      num_iter <- data_csv$sampling_info$sample_num_samples
+    }
+    percentage_divergences <- (num_of_divergences)/num_iter*100
+    message(num_of_divergences, " of ", num_iter, " (", (format(round(percentage_divergences, 0), nsmall = 0)), "%)",
+            " transitions ended with a divergence.\n",
+            "These divergent transitions indicate that HMC is not fully able to explore the posterior distribution.\n",
+            "Try increasing adapt delta closer to 1.\n",
+            "If this doesn't remove all divergences, try to reparameterize the model.\n")
+  }
+}
+
+check_sampler_transitions_treedepth <- function(data_csv) {
+  max_treedepth <- data_csv$sampling_info$sample_adapt_hmc_nuts_max_depth
+  treedepth <- posterior::extract_one_variable_matrix(data_csv$sampler, "treedepth__")
+  max_treedepth_hit <- sum(treedepth >= max_treedepth)
+    if (max_treedepth_hit > 0) {
+      if(data_csv$sampling_info$sample_save_warmup == 1) {
+      num_iter <- data_csv$sampling_info$sample_num_samples + data_csv$sampling_info$sample_num_warmup
+    } else {
+      num_iter <- data_csv$sampling_info$sample_num_samples
+    }
+    percentage_max_treedepth <- (max_treedepth_hit)/num_iter*100
+    message(max_treedepth_hit, " of ", num_iter, " (", (format(round(percentage_max_treedepth, 0), nsmall = 0)), "%)",
+            " transitions hit the maximum treedepth limit of ", max_treedepth,
+            " or 2^", max_treedepth, " leapfrog steps.\n",
```

Comment:
2^max_treedepth - 1 (I think 1023 is the max number of steps for max_treedepth = 10 )

Reply:
Ah, I was following what cmdstan diagnose outputs, but yeah, you are right.